### DISCUSSION TITLE
Can I get a contract to vote on proposals via snapshot.js

### DISCUSSION THREAD

**Discussion by builtbyleo** [snapshot-labs/snapshot#2949] [Q&A]
Hi everyone,

I'm a very new solidity developer and I was wondering if it would be possible for a contract to vote on a snapshot proposal. I know that only EOAs can sign transactions, however I wanted to see if there was a suitable work around? (For example am i able to delegate my vote to a contract). Ideally the voting power is taken from this contract rather than the actual user.

Thank you

**Comment by carlosnoesaavedra** [comment]
Good question. I'd be interested in hearing about this too

**Reply by Vipalone** [reply]
tôi cũng thấy câu hỏi rất xác thực, ai có ý kiến thì bình luận bên dưới nhé

**Comment by metrox-eth** [comment]
Yes, the way it work is that a contract can vote if it implement EIP-1271. On Snapshot a vote is a signed message, to verify a vote from a contract we call the function "isValidSignature" from that contract with the hash of the message and the signature. If that function return true then the vote is valid and can be casted / visible on Snapshot.

**Reply by builtbyleo** [reply]
thank you so much for you reply! In terms of implementation im still confused what i should put as these two parameters (web3 and account)
![image](https://user-images.githubusercontent.com/106159231/184657193-4eec407a-5596-4bf3-a541-c69a8ae199ec.png)

**Reply by metrox-eth** [reply]
@ChaituVR Can you take from here ?

**Reply by ChaituVR** [reply]
Hey @Leonard-Pat web3 and account parameters are already defined in the above code, check those two variables. :)

**Reply by builtbyleo** [reply]
Hi yes, sorry i meant if a contract were to vote instead of an EOA, could it still be done using snapshot.js api and if so how ? (would web3 / account be that of a contract instead), if not how would one do it?

**Reply by ChaituVR** [reply]
@Leonard-Pat make sure your contract support EIP1271 to be able to vote with a contract

https://eips.ethereum.org/EIPS/eip-1271

**Reply by gzliudan** [reply]
If my contract supports EIP-1271, how to test it on web browser with MetaMask ?

```solidity
contract Vote is IERC1271, Ownable {
    function isValidSignature(bytes32 hash, bytes memory signature) public view override returns (bytes4 magicValue) {
        return ECDSA.recover(hash, signature) == owner() ? this.isValidSignature.selector : bytes4(0);
    }
}
```

**Reply by BobJamesss** [reply]
Hi everyone. Once EIP-1271 is supported, how does the execution of the vote take place? What trigger's snapshots to call the isValidSignature view function on the contract?

**Comment by Turupawn** [comment]
@ChaituVR 
I have the same problem, I implemented EIP 1271 and it's not working for me

This is my contract with EIP 1271 implemented:
https://optimistic.etherscan.io/address/0xcf9eea9e11f7ad18a14a8670C757f09043A524Bf#code

This is my core that is using snapshot.org js library

```html
<h1>EIP 1271 snaptshot test</h1>

<input type="button" onclick="voteAsync()" value="vote!"></input>

<script src="https://cdn.ethers.io/lib/ethers-5.2.umd.min.js"
        type="application/javascript"></script>
<script src="https://cdn.jsdelivr.net/npm/@snapshot-labs/snapshot.js"></script>
<script>
async function voteAsync() {
    const provider = new ethers.providers.Web3Provider(window.ethereum)
    await provider.send("eth_requestAccounts", []);
    const signer = provider.getSigner()

    const hub = 'https://hub.snapshot.org'; // or https://testnet.snapshot.org for testnet
    const client = new snapshot.Client712(hub);

    const web3 = provider;
    // Smart Contract with EIP 1271 on Optimism
    const account = "0xcf9eea9e11f7ad18a14a8670C757f09043A524Bf"
    const receipt = await client.vote(web3, account, {
        space: 'opcollective.eth',
        proposal: '0x88583c43b196ec86cee45345611b582108f1d6933ab688a7cae992a6baa552a6',
        type: 'approval-voting',
        choice: 1,
        reason: 'Choice 1 make lot of sense',
        app: 'my-app'
    });
}
</script>
```

**Reply by QuixThe2nd** [reply]
I have the same problem. I've implemented EIP 1271 on my arbitrum contract but the API says signature invalid. I can't seem to find any docs/demos for this.

**Comment by Katarzynamm** [comment]
HEJ. JEST TAM KTOS LUDZKI??? PROSZĘ O KONTAKT

niedz., 1 paź 2023, 02:01 użytkownik Ahmed Castro ***@***.***>
napisał:
